src/memmachine/semantic_memory/storage/storage_base.py

```diff
@@ -0,0 +1,209 @@
+        feature: Optional[str] = None,
+        value: Optional[str] = None,
+        tag: Optional[str] = None,
+        embedding: Optional[InstanceOf[np.ndarray]] = None,
+        metadata: dict[str, Any] | None = None,
+    ):
+        raise NotImplementedError
+
+    @abstractmethod
+    async def delete_features(self, feature_ids: list[int]):
+        raise NotImplementedError
+
+    @dataclass
+    class VectorSearchOpts:
+        """Parameters controlling vector similarity constraints for retrieval."""
+
+        query_embedding: InstanceOf[np.ndarray]
+        min_distance: Optional[float] = 0.7
+
+    @abstractmethod
+    async def get_feature_set(
+        self,
+        *,
+        set_ids: Optional[list[str]] = None,
+        type_names: Optional[list[str]] = None,
+        feature_names: Optional[list[str]] = None,
+        tags: Optional[list[str]] = None,
+        k: Optional[int] = None,
+        vector_search_opts: Optional[VectorSearchOpts] = None,
+        tag_threshold: Optional[int] = None,
+        load_citations: bool = False,
+    ) -> list[SemanticFeature]:
+        """
+        Get feature set by user id
+        Return: A list of KV for each feature and value.
+           The value is an array with: feature value, feature tag and deleted, update time, create time and delete time.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    async def delete_feature_set(
+        self,
+        *,
+        set_ids: Optional[list[str]] = None,
+        type_names: Optional[list[str]] = None,
+        feature_names: Optional[list[str]] = None,
+        tags: Optional[list[str]] = None,
+        thresh: Optional[int] = None,
+        k: Optional[int] = None,
+        vector_search_opts: Optional[VectorSearchOpts] = None,
+    ):
+        """
+        Delete all the features by id
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    async def add_citations(self, feature_id: int, history_ids: list[int]):
+        raise NotImplementedError
+
+    @abstractmethod
+    async def add_history(
+        self,
+        content: str,
+        metadata: Optional[dict[str, str]] = None,
+        created_at: Optional[AwareDatetime] = None,
+    ) -> int:
+        raise NotImplementedError
+
+    @abstractmethod
+    async def get_history(
+        self,
+        history_id: int,
+    ) -> Optional[HistoryMessage]:
+        raise NotImplementedError
+
+    @abstractmethod
+    async def delete_history(
+        self,
+        history_ids: list[int],
+    ):
+        raise NotImplementedError
+
+    @abstractmethod
+    async def delete_history_messages(
+        self,
+        *,
+        start_time: Optional[AwareDatetime] = None,
+        end_time: Optional[AwareDatetime] = None,
+    ):
+        raise NotImplementedError
+
+    @abstractmethod
+    async def get_history_messages(
+        self,
+        *,
+        set_ids: Optional[list[str]] = None,
+        k: Optional[int] = None,
```

Comment:
k is not meaningful.
```suggestion
        limit: Optional[int] = None,
```